content/blog/2022-03-22-type-alias/index.md

```diff
@@ -0,0 +1,210 @@
+
+## Type-Based Alias Analysis
+TBAA operates on the program AST instead of the IR. Thus, it has access to higher level information than some other program analyses. Let's assume the language we're working with has the following kinds of memory references:
+
+1. `a.x`  Class field access
+2. `a[n]` Array indexing
+3. `*a` Pointer indirection
+
+An access path is defined to be any combination of one or more of these memory references. For instance, `(*(a.b).c[3])[2][*d[*e.f]]` is a pathological example of an access path. Basically, access paths are succinct representations of chains of memory references in the AST. We will also define typeof ($\mathcal{P}$) to be the type of the path $\mathcal{P}$.
+
+### Type Declarations Only
+
+To predict whether two paths $\mathcal{P}_1, \mathcal{P}_2$ might alias, an obvious heuristic is to say this is when the (typeof $\mathcal{P}$) has nonempty intersection with the subtypes of (typeof $\mathcal{P}$). Of course, if the two types are disjoint, then if any expression involving $\mathcal{P}_1$ type checks, the same expression with $\mathcal{P}_2$ substituted in place cannot type check, and so the two paths cannot possibly alias. We will define a function TD, which takes two access paths and returns true iff their types have a common subtype.
+
+### With Field Access
+We can extend the above heuristic by taking into account the language fact that `a.f` and `a.g` cannot alias each other for some object `a`. Here we also assume that a field access and an array indexing never alias. This is probably true for many OOP languages. We can summarize whether two access paths may alias inductively using the following table, where "FTD" is true iff its arguments paths may alias. 
+
+<style type="text/css">
+.tg  {border-collapse:collapse;border-spacing:0;margin:0px auto;}
+.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
+  overflow:hidden;padding:10px 5px;word-break:normal;}
+.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;
+  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}
+.tg .tg-a0yd{border-color:inherit;font-family:inherit;font-size:12px;text-align:center;vertical-align:top}
+.tg .tg-c3ow{border-color:inherit;text-align:center;vertical-align:top}
+</style>
+<table class="tg">
+<thead>
+  <tr>
+    <th class="tg-a0yd">$\mathcal{P}_1$</th>
+    <th class="tg-c3ow">$\mathcal{P}_2$</th>
+    <th class="tg-c3ow">FTD($\mathcal{P}_1$, $\mathcal{P}_2$)</th>
+  </tr>
+</thead>
+<tbody>
+  <tr>
+    <td class="tg-c3ow">p</td>
+    <td class="tg-c3ow">p</td>
+    <td class="tg-c3ow">true</td>
+  </tr>
+  <tr>
+    <td class="tg-c3ow">p.f</td>
+    <td class="tg-c3ow">q.g</td>
+    <td class="tg-c3ow">f=g $\land$ TD(p,q)</td>
```

Comment:
Maybe things like this would look slightly better if the whole thing were in math mode? Like `$\mathrm{f} = \mathrm{g} \land TD(\mathrm{p}, \mathrm{q})$` or similar.